runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java

```diff
@@ -3081,6 +3146,283 @@ public void testActiveWorkRefresh() throws Exception {
     assertThat(server.numGetDataRequests(), greaterThan(0));
   }
 
+  static class FakeClock implements Supplier<Instant> {
+    private static final FakeClock DEFAULT = new FakeClock();
+
+    private Instant now = Instant.now();
+
+    @Override
+    public synchronized Instant get() {
+      return now;
+    }
+
+    synchronized void sleep(Duration duration) {
+      this.now = this.now.plus(duration);
+    }
+  }
+
+  @Test
+  public void testLatencyAttributionProtobufsPopulated() throws Exception {
+    StreamingDataflowWorker.Work work =
+        new StreamingDataflowWorker.Work(null, FakeClock.DEFAULT) {
+          @Override
+          public void run() {}
+        };
+
+    FakeClock.DEFAULT.sleep(Duration.millis(10));
+    work.setState(StreamingDataflowWorker.Work.State.PROCESSING);
+    FakeClock.DEFAULT.sleep(Duration.millis(20));
+    work.setState(StreamingDataflowWorker.Work.State.COMMIT_QUEUED);
+    FakeClock.DEFAULT.sleep(Duration.millis(30));
```

Comment:
might as well test READING and COMMITTING (to test merge)

also go back to a previous state to verify that sums correctly

Reply:
Done